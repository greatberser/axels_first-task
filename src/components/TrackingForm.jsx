import { Box, Button, TextField, MenuItem } from '@mui/material';
import { Formik, Form } from 'formik';
import * as Yup from 'yup';
const categories = [
  'Food',
  'Transport',
  'Entertainment',
  'Utilities',
  'Health',
];

const validationSchema = Yup.object().shape({
  category: Yup.string().required('Category is required'),
  amount: Yup.number()
    .required('Amount is required')
    .positive('Amount must be a positive number'),
  date: Yup.date()
    .max(new Date(), 'Date cannot be in the future')
    .required('Date is required'),
});

export const TrackingForm = () => {
  console.log('TrackingForm rendered');
  return (
    <Formik
      initialValues={{ category: '', amount: '', date: '' }}
      validationSchema={validationSchema}
      onSubmit={(values, action) => {
        console.log('Saved:', values);
        action.resetForm();
      }}
    >
      {({
        handleSubmit,
        handleChange,
        values,
        errors,
        touched,
        handleBlur,
      }) => (
        <Box
          component="form"
          onSubmit={handleSubmit}
          sx={{
            display: 'flex',
            flexDirection: 'column',
            gap: 2,
            maxWidth: 400,
            margin: 'auto',
            padding: 2,
          }}
        >
          <TextField
            name="date"
            label="Date"
            value={values.date}
            onChange={handleChange}
            onBlur={handleBlur}
            error={touched.date && Boolean(errors.date)}
            helperText={touched.date && errors.date}
            type="date"
            required
          />
          {/* <TextField
            label="Date"
            type="date"
            variant="outlined"
            size="small"
            InputLabelProps={{
              shrink: true,
            }}
          /> */}

          <TextField
            select
            name="category"
            label="Category"
            value={values.category}
            onChange={handleChange}
            error={touched.category && Boolean(errors.category)}
            helperText={touched.category && errors.category}
            required
          >
            {categories.map((category) => (
              <MenuItem key={category} value={category}>
                {category}
              </MenuItem>
            ))}
          </TextField>
          {/* <TextField label="Category" select variant="outlined" size="small">
            {['Food', 'Transport', 'Entertainment', 'Utilities', 'Health'].map(
              (category) => (
                <MenuItem key={category} value={category}>
                  {category}
                </MenuItem>
              )
            )}
          </TextField> */}
          <TextField
            name="amount"
            label="Amount"
            type="number"
            value={values.amount}
            onChange={handleChange}
            onBlur={handleBlur}
            error={touched.amount && Boolean(errors.amount)}
            helperText={touched.amount && errors.amount}
            required
          />
          {/* <TextField
            label="Amount"
            type="number"
            variant="outlined"
            size="small"
          /> */}
          <Button variant="contained" color="primary">
            Add Expense
          </Button>
        </Box>
      )}
    </Formik>
  );
};

export default TrackingForm;
