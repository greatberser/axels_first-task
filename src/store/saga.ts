import { all, put, takeEvery } from 'redux-saga/effects';

function apiAddExpenses(data: any) {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve(data);
    }, 1000);
  });
}

function* sagaAddExpenses(action: any) {
  console.log('sagaAddExpenses', action.payload);
  yield put({ type: 'ADD_EXPENSE_SUCCESS', payload: action.payload });
}

export function* watchAddExpenses() {
  yield takeEvery('ADD_EXPENSES_REQUEST', sagaAddExpenses);
}

export default function* rootSaga() {
  yield all([watchAddExpenses()]);
}
