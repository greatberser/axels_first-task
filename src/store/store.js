import { configureStore } from '@reduxjs/toolkit';
import createSagaMiddleware from '@redux-saga/core';
import rootSaga from './saga';
import expensesReducer from './ducks/expenses';

const sagaMiddleware = createSagaMiddleware();

const store = configureStore({
  reducer: {
    expenses: expensesReducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(sagaMiddleware),
});

sagaMiddleware.run(rootSaga);

export default store;
