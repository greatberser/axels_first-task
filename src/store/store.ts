import { configureStore } from '@reduxjs/toolkit';
import createSagaMiddleware from '@redux-saga/core';
// @ts-ignore
import rootSaga from './saga';
import expensesReducer from './ducks/expenses';

const sagaMiddleware = createSagaMiddleware();

export const store = configureStore({
  reducer: {
    expenses: expensesReducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(sagaMiddleware),
});

sagaMiddleware.run(rootSaga);

// Типи для useSelector та useDispatch
export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
