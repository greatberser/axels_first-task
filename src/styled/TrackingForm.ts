import styled from 'styled-components';
import { Box, TextField } from '@mui/material';

export const InputContainer = styled(Box)`
  display: flex;
  flex-direction: column;
  gap: 16px;
  max-width: 400px;
  margin: auto;
  padding: 16px;
`;

export const StyledTextField = styled(TextField)`
  label {
    display: none;
  }
`;
