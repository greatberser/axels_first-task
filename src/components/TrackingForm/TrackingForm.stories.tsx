import TrackingForm from './TrackingForm';
import { Meta, StoryObj } from '@storybook/react';
import { ApolloProvider, ApolloClient, InMemoryCache } from '@apollo/client';

const client = new ApolloClient({
  uri: '/graphql',
  cache: new InMemoryCache(),
});

const meta: Meta<typeof TrackingForm> = {
  title: 'Components/TrackingForm',
  component: TrackingForm,
  tags: ['autodocs'],
  decorators: [
    (Story) => (
      <ApolloProvider client={client}>
        <Story />
      </ApolloProvider>
    ),
  ],
};

export default meta;

export const Default: StoryObj = {
  render: () => <TrackingForm />,
};
