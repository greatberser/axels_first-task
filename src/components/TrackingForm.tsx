import { Button, TextField, MenuItem } from '@mui/material';
import { Formik, FormikHelpers } from 'formik';
import * as Yup from 'yup';
import { useMutation } from '@apollo/client';

import { ADD_EXPENSE } from '../graphql/queries';

import { InputContainer, StyledTextField } from '../styled/TrackingForm';

const categories = [
  'Food',
  'Transport',
  'Entertainment',
  'Utilities',
  'Health',
];

type FormValues = {
  category: string;
  amount: number | '';
  date: string;
};

const validationSchema = Yup.object().shape({
  category: Yup.string().required('Category is required'),
  amount: Yup.number()
    .required('Amount is required')
    .positive('Amount must be a positive number'),
  date: Yup.date()
    .max(new Date(), 'Date cannot be in the future')
    .required('Date is required'),
});

const TrackingForm: React.FC = () => {
  const [addExpense] = useMutation(ADD_EXPENSE);

  const onSubmit = (values: FormValues, actions: FormikHelpers<FormValues>) => {
    try {
      addExpense({
        variables: {
          date: values.date,
          category: values.category,
          amount: parseFloat(values.amount.toString()),
        },
      });
      actions.resetForm();
    } catch (error) {
      console.error('Error adding expense:', error);
      actions.setSubmitting(false);
    }
  };

  return (
    <Formik<FormValues>
      initialValues={{ category: '', amount: '', date: '' }}
      validationSchema={validationSchema}
      onSubmit={onSubmit}
    >
      {({
        handleSubmit,
        handleChange,
        values,
        errors,
        touched,
        handleBlur,
      }) => (
        <InputContainer as="form" onSubmit={handleSubmit}>
          <TextField
            name="date"
            id="date"
            type="date"
            label="Date"
            InputLabelProps={{
              shrink: true,
            }}
            value={values.date}
            onChange={handleChange}
            onBlur={handleBlur}
            error={touched.date && Boolean(errors.date)}
            helperText={touched.date && errors.date}
          />
          <TextField
            select
            name="category"
            label="Category"
            value={values.category}
            onChange={handleChange}
            error={touched.category && Boolean(errors.category)}
            helperText={touched.category && errors.category}
          >
            {categories.map((category) => (
              <MenuItem key={category} value={category}>
                {category}
              </MenuItem>
            ))}
          </TextField>
          <TextField
            name="amount"
            label="Amount"
            type="number"
            value={values.amount}
            onChange={handleChange}
            onBlur={handleBlur}
            error={touched.amount && Boolean(errors.amount)}
            helperText={touched.amount && errors.amount}
          />
          <Button type="submit">Add Expense</Button>
        </InputContainer>
      )}
    </Formik>
  );
};

export default TrackingForm;
