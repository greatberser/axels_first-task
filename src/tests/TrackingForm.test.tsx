import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import TrackingForm from '../components/TrackingForm';
import userEvent from '@testing-library/user-event';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../store/store';

describe('TrackingForm', () => {
  it('should match a snapshot', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <MemoryRouter>
          <TrackingForm />
        </MemoryRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  describe('when form is submitted with empty fields', () => {
    beforeEach(async () => {
      render(
        <Provider store={store}>
          <MemoryRouter>
            <TrackingForm />
          </MemoryRouter>
        </Provider>
      );

      const submitButton = screen.getByRole('button', { name: /add expense/i });
      await userEvent.click(submitButton);
    });

    it('should show error for empty category', async () => {
      expect(
        await screen.findByText(/Category is required/i)
      ).toBeInTheDocument();
    });

    it('should show error for empty amount', async () => {
      expect(
        await screen.findByText(/Amount is required/i)
      ).toBeInTheDocument();
    });

    it('should show error for empty date', async () => {
      expect(await screen.findByText(/Date is required/i)).toBeInTheDocument();
    });
  });

  it('should shows error for negative amount', async () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <TrackingForm />
        </MemoryRouter>
      </Provider>
    );

    await userEvent.type(screen.getByLabelText(/date/i), '2025-07-02');
    await userEvent.click(screen.getByLabelText(/category/i));
    await userEvent.click(screen.getByRole('option', { name: /Food/i }));
    await userEvent.type(screen.getByLabelText(/amount/i), '-100');

    await userEvent.click(screen.getByRole('button', { name: /add expense/i }));

    expect(
      await screen.findByText(/Amount must be a positive number/i)
    ).toBeInTheDocument();
  });
  it('should shows error for invalid date format', async () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <TrackingForm />
        </MemoryRouter>
      </Provider>
    );

    await userEvent.type(screen.getByLabelText(/date/i), '2025-10-01');
    await userEvent.click(screen.getByLabelText(/category/i));
    await userEvent.click(screen.getByRole('option', { name: /Food/i }));
    await userEvent.type(screen.getByLabelText(/amount/i), '100');

    await userEvent.click(screen.getByRole('button', { name: /add expense/i }));

    expect(
      await screen.findByText(/Date cannot be in the future/i)
    ).toBeInTheDocument();
  });
});
