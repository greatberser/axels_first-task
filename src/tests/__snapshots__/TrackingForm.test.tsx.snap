// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`TrackingForm should match a snapshot 1`] = `
<DocumentFragment>
  <form
    class="sc-bRKDuR itjdjq"
  >
    <div
      class="MuiFormControl-root MuiTextField-root sc-hvigdm cDiwQx css-1xp5r68-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="date"
        id="date-label"
      >
        Date
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
          id="date"
          name="date"
          type="date"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-18p5xg2-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-1n64csd-MuiNotchedOutlined-root"
          >
            <span>
              Date
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root css-1xp5r68-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="«r1»"
        id="«r1»-label"
      >
        Category
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiSelect-root css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="«r1»-label «r1»"
          class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-18jp67o-MuiNativeSelect-root-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
          id="«r1»"
          role="combobox"
          tabindex="0"
        >
          <span
            aria-hidden="true"
            class="notranslate"
          >
            ​
          </span>
        </div>
        <input
          aria-hidden="true"
          aria-invalid="false"
          class="MuiSelect-nativeInput css-j0riat-MuiSelect-nativeInput"
          name="category"
          tabindex="-1"
          value=""
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-1rfqz7c-MuiSvgIcon-root-MuiNativeSelect-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-18p5xg2-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-1n64csd-MuiNotchedOutlined-root"
          >
            <span>
              Category
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root css-1xp5r68-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for="«r3»"
        id="«r3»-label"
      >
        Amount
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
          id="«r3»"
          name="amount"
          type="number"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-18p5xg2-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-1n64csd-MuiNotchedOutlined-root"
          >
            <span>
              Amount
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1uent87-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="submit"
    >
      Add Expense
    </button>
  </form>
</DocumentFragment>
`;
